### DISCUSSION TITLE
pkg-config include directories

### DISCUSSION THREAD

**Discussion by bradbell** [coin-or/Ipopt#467] [General]
The CppAD include commands include the `coin-or` directory; e.g., 
```
# include <coin-or/IpIpoptApplication.hpp>
```
But when I run
```
pkg-config ipopt --cflags
```
`/coin-or` is included at the end of the include direcotry; i.e., the include directory is the prefix following by `include/coin-or` instead of the prefix followed by `include`.

**Comment by svigerske** [comment]
So CppAD should use
```
# include "IpIpoptApplication.hpp"
```
That's why the cflags returned by pkg-config tell you add `@includedir@/coin-or` to the compilers header search path.

**Comment by bradbell** [comment]
Is not the point of putting it in a coin-or sub directory to avoid conflicts with other include files. Keeping the `coin-or` in the include command further serves this purpose.

**Reply by svigerske** [reply]
Might be. If Ipopt 3 were new, I would putting everything into an `ipopt` subdir and skip the `Ip` at the beginning of every filename.

But Ipopt 3 has been around for long, the `Ip` in the beginning is what is meant to avoid conflicts, and users (except for CppAD, apparently) are used not to use the `coin-or` prefix and to get the corresponding `-I` flag from the .pc file. I don't want to break that.

**Comment by bradbell** [comment]
I see the point for Ipopt, but going forward with a new project I think it is better to have an include file called
```
Ip/IpoptApplication.hpp
```
i.e., to group include files for a project into a directory, instead of having the same beginning in the name of every include file for a project.
